src/modm/driver/display/st7586s_impl.hpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2021, Tomasz Wasilczyk
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+#ifndef MODM_ST7586S_HPP
+#error "Do not include this file directly. Include st7586s.hpp instead."
+#endif
+
+#include <modm/architecture/interface/assert.hpp>
+#include <modm/architecture/interface/gpio.hpp>
+
+namespace modm
+{
+
+template <typename SPI, typename CS, typename RST, typename DC, int Width, int Height>
+void
+St7586s<SPI, CS, RST, DC, Width, Height>::sendCommand(Command cmd, const void *data, size_t len)
+{
+	CS::reset();
+	DC::reset();  // command mode
+	SPI::transferBlocking(static_cast<uint8_t>(cmd));
+	DC::set();  // data mode
+	if (len > 0) {
+		SPI::transferBlocking(reinterpret_cast<const uint8_t*>(data), nullptr, len);
+	}
+	CS::set();
+	// exit with data mode on
+}
+
+template <typename SPI, typename CS, typename RST, typename DC, int Width, int Height>
+void
+St7586s<SPI, CS, RST, DC, Width, Height>::initialize()
+{
+	namespace payload = impl::st7586s::payload;
+
+	// Configure GPIO
+	RST::reset();
+	RST::setOutput();
+	CS::set();
+	CS::setOutput();
+	DC::setOutput();
+
+	// Reset display
+	modm::delay_us(10);
+	RST::set();
+	modm::delay_ms(120);
+
+	// Power ON operation flow (see datasheet)
+	sendCommand(Command::SleepOff);
+	sendCommand(Command::DisplayOff);
+	modm::delay_ms(50);  // t_{ON-V2}
+	sendCommand(Command::SetVop, payload::SetVop(13.52f));
+	sendCommand(Command::SetBias, payload::SetBias::Ratio_1_11);
+	sendCommand(Command::SetBooster, payload::SetBooster::x8);
+	sendCommand(Command::AnalogControl, payload::AnalogControl::Enable);
+	sendCommand<payload::NLineInversion_t>(Command::NLineInversion,
+		payload::LineInversionType_t(payload::LineInversionType::FrameInversion));
+	sendCommand(Command::DisplayModeMono);  // TODO: grayscale
+	sendCommand(Command::EnableDdram, payload::EnableDdram::Enable);
+	sendCommand(Command::ScanDirection, payload::DisplayControl::ComInc
+		| payload::DisplayControl::SegInc);
+	sendCommand(Command::DisplayDuty, uint8_t(Height - 1));
+	sendCommand(Command::InverseOff);
+	setClipping(0, 0, Width, Height);
+	update();  // clear the display
+	sendCommand(Command::DisplayOn);
+}
+
+template <typename SPI, typename CS, typename RST, typename DC, int Width, int Height>
+void
+St7586s<SPI, CS, RST, DC, Width, Height>::setClipping(uint16_t x, uint16_t y, uint16_t w, uint16_t h)
+{
+	namespace payload = impl::st7586s::payload;
```

Comment:
You can drop a sole `using payload = impl::st586s::payload` within class declaration to spare repetitions.

Reply:
Ditto.